### ISSUE TITLE
[BUG] SireError::invalid_cast when merging a ligand with itself

### ISSUE THREAD

**Issue by msuruzhon** [OpenBioSim/biosimspace#23]
Hi,

I have an issue with some [ligands](https://github.com/OpenBioSim/biosimspace/files/10958704/ligands.sdf.zip) from the OpenFF Benchmark set (so apologies I haven't reproduced it with a faster to parametrise system). The error I get when merging them is the following:

```
TypeError: SireError::invalid_cast: Cannot cast from an object of class 
"SireMol::AtomPropertyProperty" to an object of class "SireMol::AtomCharges". 
(call sire.error.get_last_error_details() for more info)
```

```python
import BioSimSpace as BSS

lig = BSS.IO.readMolecules("ligands.sdf")[0]
lig = BSS.Parameters.gaff2(lig).getMolecule()
merged_lig = BSS.Align.merge(lig, lig)
```

It seems to only happen with some ligands so it is probably something related to the input SDF file.

Many thanks.

**Comment by lohedges** [comment]
It's failing trying to assign the first (and only) non-zero formal charge in the lambda = 0 state. Looking at the property keys of the ligand, the formal charge is as follows:

```python
lig._sire_object.property("formal_charge")
Out[11]:
SireMol::AtomCharges( size=55
0: 0 |e|
1: 0 |e|
2: 0 |e|
3: 0 |e|
4: 0 |e|
...
50: 0 |e|
51: 0 |e|
52: 0 |e|
53: 0 |e|
54: 0 |e|
)
```

It _is_ an `AtomCharges` property, so I'm confused by the message in the exception. Simply removing the `formal_charge` property makes things work:

```python
lig._sire_object = lig._sire_object.edit().removeProperty("formal_charge").molecule().commit()
```

I'll take a look at the SDF parser to see how `formal_charge` is being handled. It's not needed for the merge, so we should just be able to ignore it if it is proving to be problematic.

**Comment by lohedges** [comment]
Okay, I think I've figured it out. Now in Sire, properties that are zero a strictly unitless when returned, e.g.:

```python
print(lig.getAtoms()[37]._sire_object.property("formal_charge"))
0
print(lig.getAtoms()[37]._sire_object.property("formal_charge").CHARGE())
0

print(lig.getAtoms()[38]._sire_object.property("formal_charge"))
1 |e|
print(lig.getAtoms()[38]._sire_object.property("formal_charge").CHARGE())
1
```

This means that when the formal charges at lambda = 0 are reassigned, the property container is initially set to a generic `AtomPropertyProperty` since the formal charge is dimensionless. When we later try to assign a non-zero value it complains that the type doesn't match the existing container.

This is a pain, since I imagine that we will have to check for zero values for _any_ property that has types in the merge code and elsewhere. @chryswoods: Is there an easy workaround for this? Here I am just looping over arbitrary properties, so I don't really want to have to do any work to have to infer the type, or work it out from the property name, or search for non-zero values of the property in order to figure out what it is.

**Comment by lohedges** [comment]
I'm hoping formal charge is an edge case, since it's clear when values are strictly zero.

**Comment by chryswoods** [comment]
In these cases it is better to set the charge initially to a non-zero value (so that sire knows it is a charge) and then set it to zero. In C++-land the typing is explicit, so setting 0 works (you use `.setProperty<Charge>(0)`). Python misses this.

One solution may be to add a `set_property` function that takes the unit type as a second argument? This could internally set the property to `1 * unit` and then set the property to the desired value. For example

```
def _set_property(obj, key, value, unit=None):
    if unit is not None:
        obj.__orig__set_property(key, 1*unit)

    obj.__orig__set_property(key, value)
```

I am surprised that this has been lost though, as the `AtomCharges` are set correctly in the SDF parser. The `AtomCharges` object [is created explicitly ](https://github.com/OpenBioSim/sire/blob/f5955c51bcb1a2f23d9b66a6d640d2c71380546d/corelib/src/libs/SireIO/sdf.cpp#L1808), it is populated with values that are [explicitly charge values](https://github.com/OpenBioSim/sire/blob/f5955c51bcb1a2f23d9b66a6d640d2c71380546d/corelib/src/libs/SireIO/sdf.cpp#L1825) and the charge property is [being set correctly](https://github.com/OpenBioSim/sire/blob/f5955c51bcb1a2f23d9b66a6d640d2c71380546d/corelib/src/libs/SireIO/sdf.cpp#L1885).

Ok - I see the error is that BioSimSpace copies out the charges and reassigns them. I think that the solution is that BioSimSpace should copy the property as a whole across first, and then set the individual atom properties, e.g.

```
oldmol = source of properties
newmol = molecule into which properties will be copied

newmol = newmol.edit().set_property("charge", oldmol.property("charge_at_lambda0")).commit()
```

This should work because the property is really an `AtomCharges`, even though some of the charges are zero.

**Comment by lohedges** [comment]
Ah, interesting thanks. The reason that I do it atom by atom is because I'm building a new molecule to represent the merged topology at each end state. I can't just copy the property from one molecule to another because it won't necessarily be the same size, i.e. it depends on the mapping. One end state may have additional atoms (dummies) so the property containers will be the wrong size for copying. Trying to copy will fail, since the molecular layout is incompatible. (In this case it would work, since we are merging a ligand with itself.)

**Comment by lohedges** [comment]
Thinking about it, I can probably just call `.toVector()` on the property, add the extra ones that I need, then assign as a whole. Will try that.

**Comment by msuruzhon** [comment]
> I'm hoping formal charge is an edge case, since it's clear when values are strictly zero.

I also saw this issue couple of days ago and I believe that this time the property in question was `AtomicRadii` so I don't think it is specific to charges.

**Comment by lohedges** [comment]
Thanks, that's good to know.

**Comment by chryswoods** [comment]
I think one way to solve this is to create an empty `AtomProperty<T>` for the property you are creating, putting the parameters into a list, and then copying the parameters into the `AtomProperty<T>` in one go (and then assigning that property to the molecule). This may be faster.

Here's how to create a new `AtomProperty<T>` for a molecule.

```
# get the charges from the old molecule
old_chgs = oldmol.property("charge")

# create an empty AtomCharges for the new molecule 
# (this possibly has a different layout)
new_chgs = type(old_chgs)(newmol.info())

# Note how above we have called the constructor of the 
# property object directly, passing in a molecule info so
# that it can size itself for the molecule. You can 
# check this by printing it - is should be a set of 
# zero charges for all atoms
print(new_chgs)

# Now get the list of all charges from the old molecule
chgs_list = oldchgs.to_vector()

# the list is in AtomIdx order. Change the values as you 
# need, inserting new charges for added atoms or deleting
# charges for removed atoms. Once this is done, the chgs_list
# should contain the new charges in AtomIdx order for the 
# new molecule

# now assign the charges into the new charge property
new_chgs.copy_from(chgs_list)

# You can check these are correct, e.g.
print(new_chgs)

# now set this as the new property in the new molecule
newmol = newmol.edit().set_property("charge", new_chgs).commit()
```

**Comment by chryswoods** [comment]
Only properties that are sub-view properties (e.g. `AtomProperty`, `ResProperty` etc) will have a `to_vector()` function. The molecule-level properties (e.g. `Connectivity`) don't have `to_vector`.

You could use `hasattr(p, "to_vector")` to identify which properties are sub-view properties and then do the mapping above for those cases.

There is also the `property_type` function that can tell you the underlying C++ type for a property, e.g.

```
>>> print(mol.property_type("charge"))
'SireMol::AtomCharges'
```

**Comment by lohedges** [comment]
Thanks, I'll look into this. I've currently got it working by checking for a zero value property then inferring the type and apply the original unit. This seems to be okay.

I agree that doing things in one go is probably the way forward, though. The issue is that we have a mixture of shared properties (present in both molecules, so can be perturbed) and properties that are unique to either molecule, so might not be present for all atoms in an end state. I'd need to account for this in your suggested approach, i.e. there could be empty values in the property list that would need to be given default (zero) values for that type. The current atom-by-atom approach worked for both situations and meant that I could use the exact same logic for all of the properties.

I'll have a play around and see how I get on.

**Comment by chryswoods** [comment]
You could keep the atom-by-atom approach by constructing a default-valued property and adding that to the new molecule. All of the atom values would be default (likely zero) and would then only be updated when you set them. This will prevent the error caused by adding a zero value as the first property, so sire then defaults to making all properties a AtomPropertyProperty type.

For example

```
oldchg = oldmol.property("charge")

zerochg = type(oldchg)(newmol.info())

newmol = newmol.edit().set_property("charge", zerochg).commit()
```

Now `newmol` has a charge property for all atoms which are zero, but are of the correct type (`charge`).

**Comment by chryswoods** [comment]
You could make it generic by putting it behind a try/except block?

```
oldprop = oldmol.property("property_name")

try:
    default_prop = type(oldprop)(newmol.info())
    newmol = newmol.edit().set_property("property_name", default_prop).commit()
except Exception:
    # this wasn't an AtomProperty or similar
    pass

# You can now access the property via an atom view, 
# safe in the knowledge it will have the same type as
# in `oldmol`
cursor = newmol.cursor()
cursor.atoms()[0]["property_name"] = 0.0
cursor.atoms()[1]["property_name"] = 3 * sr.units.mod_electron
newmol.commit()
```

**Comment by lohedges** [comment]
Great, that makes sense. Will try that.

**Comment by lohedges** [comment]
Unfortunately it doesn't seem to work consistently. While it solves the issue for `formal_charge`, which are `AtomCharges`, it doesn't work for the `gb_radii` property, which is of type `AtomRadii`. This is despite both properties having had a default-valued property set. (It still fails when trying to set a `gb_radii` value to zero.)

For example:

```python
import BioSImSpace as BSS

propane = BSS.Parameters.parameterise("CCC", "openff_unconstrained-2.0.0").getMolecule()

# All atoms have zero for the "gb_radii" property. Try to reset one.
atom = propane.getAtoms()[0]._sire_object
print("gb_radii", atom.property("gb_radii"))

edit_mol = propane._sire_object.edit()
edit_mol.atom(atom.index()).setProperty("gb_radii", atom.property("gb_radii"))
```

Gives:

```python
gb_radii 0
TypeError: SireError::invalid_cast: Cannot cast from an object of class "SireMol::AtomRadii" to an object of class
"SireMol::AtomPropertyProperty". (call sire.error.get_last_error_details() for more info)
```

Whereas, doing the same for a zero `AtomCharges` property works, e.g.:

```python
import BioSimSpace as BSS

lig = BSS.IO.readMolecules("ligands.sdf")[0]

# The first 37 atoms have 0 for formal_charge. Try resetting one of them.
atom = lig.getAtoms()[0]._sire_object
print("formal_charge", atom.property("formal_charge"))

edit_mol = lig._sire_object.edit()
edit_mol.atom(atom.index()).setProperty("formal_charge", atom.property("formal_charge"))
```

Gives:

```python
formal_charge 0
Editor{ Atom( H:1     [  -2.16,   14.46,   -7.77] ) }
```
In this particular case, all of the `gb_radii` for propane are zero. I'm wondering if this is the issue, even though it is of the correct type, i.e.:

```python
type(propane._sire_object.property("gb_radii"))
sire.legacy.Mol._Mol.AtomRadii

propane._sire_object.property("gb_radii").toVector()
[0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 ]
```

**Comment by lohedges** [comment]
Okay, it's because the `__init__.py` for the `sire.legacy.Mol` wrapper has no special handling for `AtomRadii`, e.g.:

```python
...
    try:
        return getattr(molview, "_set_property_%s" % typename)(key, property)
    except AttributeError as e:
        if orig_typename is not None:
            # we need to go for the original typename, if the property
            # exists already
            value = None

            if orig_typename.endswith("Charges"):
                from ..Units import mod_electron

                value = mod_electron
            elif orig_typename.endswith("Masses"):
                from ..Units import g_per_mol

                value = g_per_mol

            if value is not None:
                (typename, value) = __get_typename__(value)
                try:
                    return getattr(molview, "_set_property_%s" % typename)(
                        key, property
                    )
                except AttributeError as e:
                    print(e)

        # no matching function - see if we have the generic 'PropertyProperty'
        if hasattr(molview, "_set_property_SireBase_PropertyPtr"):
            return molview._set_property_SireBase_PropertyPtr(
                key, _Base.wrap(property)
            )

        raise e
```

You can see it deals with charges and masses, but not radii. I'll fix this and raise a PR. (This is basically what I was doing in BioSImSpace anyway, but it will be good to have this in the Sire layer.)

**Comment by chryswoods** [comment]
Thanks - yes, there's a lot of magic and kludgy code under the hood to make the property system act like Python but actually be C++ typed underneath... Thanks for sorting this out :-)

**Comment by msuruzhon** [comment]
Thank you both for fixing it!
